staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go

```diff
@@ -0,0 +1,279 @@
+//
+// cachingObject implements the metav1.Object interface (accessors for
+// all metadata fields). However, setters for all fields except from
+// SelfLink (which is set lately in the path) are ignored.
+type cachingObject struct {
+	// Object for which serializations are cached.
+	object metaRuntimeInterface
+
+	lock sync.Mutex
+	// serializations is a cache containing object`s serializations.
+	// The value stored in atomic.Value is of type serializationsCache.
+	// The atomic.Value type is used to allow fast-path.
+	serializations atomic.Value
+
+	// selfLink is the value passed in the first SetSelfLink() call.
+	selfLink *string
+}
+
+// newCachingObject returns cachingObject wrapping a given object.
+// An error is returned if it's not possible to cast the object to
+// metav1.Object type.
+func newCachingObject(object runtime.Object) (*cachingObject, error) {
+	if obj, ok := object.(metaRuntimeInterface); ok {
+		result := &cachingObject{object: obj.DeepCopyObject().(metaRuntimeInterface)}
+		result.serializations.Store(make(serializationsCache))
+		return result, nil
+	}
+	return nil, fmt.Errorf("can't cast object to metav1.Object: %#v", object)
+}
+
+// CacheEncode implements runtime.CacheableObject interface.
+// It serializes the object and writes the result to given io.Writer trying
+// to first use the already cached result and falls back to a given encode
+// function in case of cache miss.
+// It assumes that for a given identifier, the encode function always encodes
+// each input object into the same output format.
+func (o *cachingObject) CacheEncode(id runtime.Identifier, encode func(runtime.Object, io.Writer) error, w io.Writer) error {
+	result := func() *serializationResult {
+		// Fast-path for getting from cache.
+		serializations := o.serializations.Load().(serializationsCache)
+		if result, exists := serializations[id]; exists {
+			return result
+		}
+
+		// Slow-path (that may require insert).
+		o.lock.Lock()
+		defer o.lock.Unlock()
+
+		serializations = o.serializations.Load().(serializationsCache)
+		// Check if in the meantime it wasn't inserted.
+		if result, exists := serializations[id]; exists {
+			return result
+		}
+
+		// Insert an entry for <id>. This requires copy of existing map.
+		newSerializations := make(serializationsCache)
+		for k, v := range serializations {
+			newSerializations[k] = v
+		}
+		result := &serializationResult{}
+		newSerializations[id] = result
+		o.serializations.Store(newSerializations)
+		return result
+	}()
+	result.once.Do(func() {
+		buffer := bytes.NewBuffer(nil)
+		result.err = encode(o.GetObject(), buffer)
+		result.raw = buffer.Bytes()
+	})
+	// Once invoked, fields of serialization will not change.
+	if result.err != nil {
+		return result.err
+	}
+	_, err := w.Write(result.raw)
+	return err
+}
+
+// GetObject implements runtime.CacheableObject interface.
+func (o *cachingObject) GetObject() runtime.Object {
+	result := o.object.DeepCopyObject().(metaRuntimeInterface)
```

Comment:
I wasn't expecting to get a copy here... the ownership semantics of what is returned here need to be clarified in the CacheableObject interface. Is the CacheableObject implementer responsible for copying, or is the caller of GetObject() responsible for copying?

Reply:
added a comment (both here and in the interface)